spirv_hlsl.cpp

```diff
@@ -0,0 +1,1055 @@
+		// to use same name for variables.
+		// Since we want to make the GLSL debuggable and somewhat sane, use fallback names for variables which are duplicates.
+		add_local_variable_name(arg.id);
+
+		decl += argument_decl(arg);
+		if (&arg != &func.arguments.back())
+			decl += ", ";
+
+		// Hold a pointer to the parameter so we can invalidate the readonly field if needed.
+		auto *var = maybe_get<SPIRVariable>(arg.id);
+		if (var)
+			var->parameter = &arg;
+	}
+
+	decl += ")";
+	statement(decl);
+}
+
+void CompilerHLSL::emit_hlsl_entry_point()
+{
+	auto &execution = get_entry_point();
+	const char *post = "Frag";
+	if (execution.model == ExecutionModelVertex)
+	{
+		post = "Vert";
+	}
+
+	statement("Output", post, " main(Input", post, " input)");
+	begin_scope();
+
+	for (auto &id : ids)
+	{
+		if (id.get_type() == TypeVariable)
+		{
+			auto &var = id.get<SPIRVariable>();
+			auto &type = get<SPIRType>(var.basetype);
+
+			if (var.storage != StorageClassFunction && !var.remapped_variable && type.pointer &&
+			    var.storage == StorageClassInput && interface_variable_exists_in_entry_point(var.self))
+			{
+				if (execution.model == ExecutionModelVertex && is_builtin_variable(var))
+					continue;
+
+				auto &m = meta[var.self].decoration;
+				auto &mtype = get<SPIRType>(var.basetype);
+				if (mtype.vecsize == 4 && mtype.columns == 4)
+				{
+					statement(m.alias, "[0] = input.", m.alias, "_0;");
+					statement(m.alias, "[1] = input.", m.alias, "_1;");
+					statement(m.alias, "[2] = input.", m.alias, "_2;");
+					statement(m.alias, "[3] = input.", m.alias, "_3;");
+				}
+				else
+				{
+					statement(m.alias, " = input.", m.alias, ";");
+				}
+			}
+		}
+	}
+
+	if (execution.model == ExecutionModelVertex)
+	{
+		statement("vert_main();");
+	}
+	else
+	{
+		statement("frag_main();");
+	}
+
+	statement("Output", post, " output;");
+
+	for (auto &id : ids)
+	{
+		if (id.get_type() == TypeVariable)
+		{
+			auto &var = id.get<SPIRVariable>();
+			auto &type = get<SPIRType>(var.basetype);
+
+			if (var.storage != StorageClassFunction && !var.remapped_variable && type.pointer &&
+			    var.storage == StorageClassOutput && interface_variable_exists_in_entry_point(var.self))
+			{
+				auto &m = meta[var.self].decoration;
+				bool is_no_builtin = !is_builtin_variable(var) && !var.remapped_variable;
+				if (is_no_builtin)
+					statement("output.", m.alias, " = ", m.alias, ";");
+				else if (execution.model == ExecutionModelVertex)
+				{
+					statement("output.gl_Position = gl_Position;");
+				}
+			}
+		}
+	}
+
+	if (execution.model == ExecutionModelVertex)
+	{
+		if (options.shader_model <= 30)
+		{
+			statement("output.gl_Position.x = output.gl_Position.x - gl_HalfPixel.x * output.gl_Position.w;");
+			statement("output.gl_Position.y = output.gl_Position.y + gl_HalfPixel.y * output.gl_Position.w;");
+		}
+		statement("output.gl_Position.z = (output.gl_Position.z + output.gl_Position.w) * 0.5;");
```

Comment:
Kinda expected some optional Y-flipping, considering DX Y-flips clip space to get to window space.